### PR TITLE
Clean up dependency tracking in Rustbuild [2/2]

### PR DESCRIPTION
**Pull request by collin5**
Make `clear_if_dirty` calls in `Builder::cargo` with stamp dependencies for the given Mode. 

Continuation of #50904
Ref issue #50509 
r? @Mark-Simulacrum

### CODE REVIEW

**Review by Mark-Simulacrum — COMMENTED** [review]
This looks great, left some nits but should be mostly minor.

File: src/bootstrap/tool.rs [hunk 1]
```diff
@@ -38,33 +38,10 @@ impl Step for CleanTools {
         run.never()
     }
 
-    fn run(self, builder: &Builder) {
-        let compiler = self.compiler;
-        let target = self.target;
+    fn run(self, _builder: &Builder) {
```
[the remark this answers is not in the document]
Comment (collin5) [reply src/bootstrap/tool.rs]: Wasn't sure about this at first, but yeah looks like it won't needed anymore 👍

File: src/bootstrap/builder.rs [hunk 2]
```diff
@@ -705,6 +705,68 @@ impl<'a> Builder<'a> {
...
+            // This is the intended out directory for crate documentation.
+            "doc" =>  self.crate_doc_out(target),
+
+            _ => self.stage_out(compiler, mode),
+        };
+
+        // This is for the original compiler, but if we're forced to use stage 1, then
+        // std/test/rustc stamps won't exist in stage 2, so we need to get those from stage 1, since
+        // we copy the libs forward.
+        let compiler = if self.force_use_stage1(compiler, target) {
```
[the remark this answers is not in the document]
Comment (collin5) [reply src/bootstrap/builder.rs]: Right, going to fix this. Thanks!

File: src/bootstrap/builder.rs [hunk 3]
```diff
@@ -705,6 +705,68 @@ impl<'a> Builder<'a> {
     ) -> Command {
         let mut cargo = Command::new(&self.initial_cargo);
         let out_dir = self.stage_out(compiler, mode);
+
+        let mut my_out = match cmd {
```
[the remark this answers is not in the document]
Comment (collin5) [reply src/bootstrap/builder.rs]: I had thought of unifying it but looks like this (`out_dir`) is used lower down https://github.com/rust-lang/rust/pull/52036/files#diff-4bc384d35e57dc577b9378d4d733d7c3R771 and is always the same no matter the passed `cmd` yet `my_out` should be different for each passed `cmd`. Maybe I'm missing something here.

[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply src/bootstrap/builder.rs]: Ah, right -- yes, that's correct. I think we'll eventually want to resolve that to be the same (since presumably that's ultimately what we want) but for now it should be enough to just add the comment -- basically saying that this is a command-specific path, unlike out_dir which is broader in scope.

**Comment by alexcrichton** [comment]
This looks pretty good to me, thanks! 

I'd recommend being sure to try out a few combinations of commands locally, this'll be easy to accidentally introduce a regression as we don't have many workflow-like tests

**Comment by collin5** [comment]
You can now take another look at this. Thank you!
r? @Mark-Simulacrum

File: src/bootstrap/builder.rs [hunk 4]
```diff
@@ -705,6 +705,78 @@ impl<'a> Builder<'a> {
...
+                },
+                Mode::Test => {
+                    self.clear_if_dirty(&my_out, &libstd_stamp);
+                },
+                Mode::Rustc => {
+                    self.clear_if_dirty(&my_out, &self.rustc(compiler));
+                    self.clear_if_dirty(&my_out, &libstd_stamp);
+                    self.clear_if_dirty(&my_out, &libtest_stamp);
+                },
+                Mode::Codegen => { },
```
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs]: Codegen also needs to be cleared if rustc/std/test/rustc are dirty.

File: src/bootstrap/compile.rs [hunk 5]
```diff
@@ -248,11 +245,7 @@ impl Step for StdLink {
             copy_apple_sanitizer_dylibs(builder, &builder.native_dir(target), "osx", &libdir);
         }
 
-        builder.ensure(tool::CleanTools {
-            compiler: target_compiler,
-            target,
-            cause: Mode::Std,
-        });
+        builder.cargo(target_compiler, Mode::ToolStd, target, "clean");
```
Comment (Mark-Simulacrum) [inline src/bootstrap/compile.rs]: Hm, is this (and similar statements below) actually necessary? I would've hoped this would happen on its own.

File: src/bootstrap/builder.rs (hunk 4 above)
Comment (collin5) [reply src/bootstrap/builder.rs]: Going to fix this ASAP thanks!

File: src/bootstrap/compile.rs (hunk 5 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/compile.rs]: Step resolution shouldn't depend on files existing on the system - do you happen to have a concrete example? If we actually do need this then can we factor out the cleaning logic from `Builder::cargo` and call that separately here? It feels a little odd to be creating the full command and then not running it (i.e., using implicit side-effects).

Comment (Mark-Simulacrum) [reply src/bootstrap/compile.rs]: If you don't feel like dealing with this though feel free to let me know -- I'm fine with leaving these in for now

Comment (collin5) [reply src/bootstrap/compile.rs]: Yeah, `i.e` removing both https://github.com/rust-lang/rust/pull/52036/files#diff-fa9668c926a2788f7849514fe3ed66a9R248 for `StdLink` and https://github.com/rust-lang/rust/pull/52036/files#diff-fa9668c926a2788f7849514fe3ed66a9R446 for `TestLink` then doing `./x.py test src/bootstrap`, Assertion https://github.com/rust-lang/rust/blob/1ecf6929dc3b309cdfcb7239260777dab38242b9/src/bootstrap/builder.rs#L1660 will fail and looking at diff shows one of steps was missed. Perhaps we could factor out cleaning logic as you had suggested.

Comment (Mark-Simulacrum) [reply src/bootstrap/compile.rs]: Okay, let's leave this as is for now and look into the details later -- it seems fine to leave this somewhat half done pending a more thorough investigation as to how *exactly* the dependency edge is created.

**Comment by Mark-Simulacrum** [comment]
@collin5 If you can try out a few workflows/commands and make sure they continue to work (i.e., no weird linker/compiler errors related to duplicate deps or unexpected rebuilding), such as:
 - `x.py build --stage 1 src/libstd` with edits to compiler and std (i.e., compiler successfully builds after such changes)
 - `x.py test --stage 1 src/test/compile-fail` with an edit to std (making sure that tools build properly)
 - `x.py build --stage 2` with edit to rustc (basically, ensure that editing rustc doesn't cause a later step to fail due to deps being outdated).

@bors delegate+

If you feel about the above "workflows" being correct then feel free to r=me, you should have permission now.

**Comment by collin5** [comment]
@Mark-Simulacrum Looks like this is failing for `--target wasm32`. Not yet sure why though.

**Comment by Mark-Simulacrum** [comment]
@collin5 Want me to take a look and see if I can help debug the issue?

**Comment by collin5** [comment]
Yeah, that would be great 👍 . Thank you.

**Comment by Mark-Simulacrum** [comment]
I've run into some difficulties, primarily caused by the current version in this PR rebuilding everything on every build. I haven't yet tracked it down fully but I suspect that the clear_if_dirty calls are writing the stamp files too aggressively in some fashion and overwriting previous iterations, causing us to rebuild unnecessarily. If you can fix that then we can take a further look at the double core issue but I suspect it might have the same underlying cause.

**Comment by collin5** [comment]
Ok let me take a look. Also, were you able to reproduce this locally? All cases work correctly if I don't specify the target and If I do `--target`, I get a different error `error: linker `rust-lld` not found`. @Mark-Simulacrum

**Comment by Mark-Simulacrum** [comment]
Yes, I'm able to reproduce locally. Not sure about the linker error, but you should be able to use the docker container (`./src/ci/docker/run.sh wasm32-unknown`) given that you have docker installed.

**Comment by collin5** [comment]
Huh! been looking for a way to do this for quite some time. Thanks a lot 👍

**Comment by TimNN** [comment]
Ping from triage, @collin5! What is the status of this PR?

**Comment by pietroalbini** [comment]
Ping from triage @collin5! It's been a while since we heard from you, will you have time to work on this again?

**Comment by collin5** [comment]
Hi @Mark-Simulacrum I've run a couple of tests and this (wasm32-unknown) looks to be fixed. Problem may have been with setting stamp `compiler` to `builder.compiler(1, compiler.host)` in `builder::cargo`. Moved this to only steps that require it. Wanted you to take a look at the changes before doing r+. Thanks!

File: src/bootstrap/compile.rs [hunk 6]
```diff
@@ -655,7 +677,6 @@ impl Step for CodegenBackend {
         }
 
         let out_dir = builder.cargo_out(compiler, Mode::Codegen, target);
-        builder.clear_if_dirty(&out_dir, &librustc_stamp(builder, compiler, target));
```
Comment (Mark-Simulacrum) [inline src/bootstrap/compile.rs]: Hm, is this change required? I'd expect that if we change rustc crates we need this in order for the codegen backends to be properly rebuilt.

**Comment by Mark-Simulacrum** [comment]
@bors r+

Looks good -- thanks so much!

**Comment by kennytm** [comment]
@bors p=111

Prepare for rollup fairness

**Comment by collin5** [comment]
Hmm, Looks like I've failed to reproduce this. Running `./src/ci/docker/run.sh wasm32-unknown` on `Ubuntu 16.04.2 LTS`  always exits with 
```
[TIMING] Crate { compiler: Compiler { stage: 2, host: "x86_64-unknown-linux-gnu" }, target: "wasm32-unknown-unknown", mode: Std, test_kind: Test, krate: "core" } -- 1.078
Build completed successfully in 0:02:33
``` 
Do you please mind taking a look at this? Maybe there is something I'm missing here. Thank you! cc @Mark-Simulacrum

**Comment by collin5** [comment]
Yeah, I did that. Even removed all the docker images and cache but still getting the same result.
Thank you, I would really appreciate that.

**Comment by collin5** [comment]
Awesome! thanks a lot. Going to pull the changes in a few.

**Comment by collin5** [comment]
:relieved: Whew! . Finally this should do it.
@bors r+
